### PR TITLE
perf(xds): use aggregated mesh context for zone proxies

### PR DESCRIPTION
**Pull request by jakubdyszkiewicz**
### Checklist prior to review

With the recent VirtualOutbound cross zone work I noticed that we can heavily use MeshContext from specific meshes. If we do, then it's very easy to avoid unnecessary zone proxies XDS generation by checking the aggregated mesh hash following the same pattern as we do for regular DP proxies.

This improvement reduces:
* Number of queries to DB, instead of doing separate DB queries to all DPPs or all MeshGateways we can just aggregate all from all mesh contexts
* Number of XDS reconciliations for zone proxies if there are no changes in the system

Note: I'm not sure if `MeshContexts` is a good name. It can be easily confused with `MeshContext` when reading the code. Should this be `AggregatedMeshContexts`? `AllMeshContexts`?

`meshContextFetcher` is introduced so we don't have dependency on mesh cache in xds/context.

- [ ] [Link to relevant issue][1] as well as docs and UI issues --
- [ ] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [ ] Tests (Unit test, E2E tests, manual test on universal and k8s) --
- [ ] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [ ] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? --
- [ ] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch) or add a `ci/` label to run fewer/more tests?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/xds/context/context.go [hunk 1]
```diff
@@ -74,3 +74,41 @@ func (mc *MeshContext) GetLoggingBackend(tl *core_mesh.TrafficLogResource) *mesh
...
+type MeshContexts struct {
+	Hash               string
+	Meshes             []*core_mesh.MeshResource
+	MeshContextsByName map[string]MeshContext
+	ZoneEgressByName   map[string]*core_mesh.ZoneEgressResource
+}
+
+func (m MeshContexts) AllDataplanes() []*core_mesh.DataplaneResource {
+	var resources []*core_mesh.DataplaneResource
+	for _, meshCtx := range m.MeshContextsByName {
```
Comment (lobkovilya) [inline pkg/xds/context/context.go:88]: Each individual MeshContext has all resources sorted. I think we should keep the order for aggregated meshContexts, so generator could rely on this.

File: pkg/xds/sync/dataplane_watchdog.go [hunk 2]
```diff
@@ -165,7 +165,25 @@ func (d *DataplaneWatchdog) syncIngress(ctx context.Context, metadata *core_xds.
 		ControlPlane: d.EnvoyCpCtx,
 		Mesh:         xds_context.MeshContext{}, // ZoneIngress does not have a mesh!
 	}
-	proxy, err := d.IngressProxyBuilder.Build(ctx, d.key)
+
+	meshContexts, err := xds_context.AggregateMeshContexts(ctx, d.ResManager, d.MeshCache.GetMeshContext)
```
Comment (lobkovilya) [inline pkg/xds/sync/dataplane_watchdog.go:169]: Should we use `d.MeshCache.GetMeshContext` to get individual MeshContexts for the aggregated?

File: pkg/xds/sync/ingress_proxy_builder.go [hunk 3]
```diff
@@ -96,74 +77,68 @@ func (p *IngressProxyBuilder) buildZoneIngressProxy(
...
 			continue
 		}
-		meshName := mesh.GetMeta().GetName()
 
-		meshCtx, err := p.meshCache.GetMeshContext(ctx, meshName)
-		if err != nil {
-			return nil, err
+		meshCtx, ok := meshContexts.MeshContextsByName[mesh.GetMeta().GetName()]
+		if !ok {
+			panic("there should be a corresponding mesh context for every mesh in mesh contexts")
```
Comment (lobkovilya) [inline pkg/xds/sync/ingress_proxy_builder.go:131]: Maybe you can add a method to MeshContexts called `MustGetMesh` that panics when there is no mesh

File: pkg/xds/sync/dataplane_watchdog.go (hunk 2 above)
Comment (jakubdyszkiewicz) [reply pkg/xds/sync/dataplane_watchdog.go:169]: what's the alternative?

**Review by lobkovilya — APPROVED** [review]

**Pull request merged**